### DISCUSSION TITLE
RFC - ESLint plugin for Next

### DISCUSSION THREAD

**Discussion by janicklas-ralph** [vercel/next.js#11731] [RFC]
# Goal
Build an Eslint plugin for NextJS that includes specific rules to help report common conformance mistakes that we’ve noticed users make during development time.

# Summary
Our initial efforts to add conformance checks to NextJs resulted in building a webpack plugin - [webpack-conformance-plugin](https://github.com/zeit/next.js/tree/canary/packages/next/build/webpack/plugins/webpack-conformance-plugin), which runs as part of `next build` and reports errors/warning  as part of the build output. 
There were two main issues with this
- Potential increase in build times, because of the added execution time of webpack-conformance-plugin
- Delayed feedback loop. Since users see the errors/warnings only when they run the next build or another command, there is potential rework needed when they have to go back and fix the issues.

An ESlint plugin would help us solve both these issues. The plugin would give development time feedback to users (even in their editors) and avoid any rework related to best practices enforcements from conformance. The plugin also offloads conformance checks from the `webpack-conformance-plugin` and moves it into ESLint plugin which will reduce the build times when we run build-time specific conformance checks.

# Approach
We identify conformance checks that can be run during development time and move the checks into the new ESLint plugin for NextJs called `eslint-plugin-next`.
We then integrate this by default in `create-next-app` and also make community suggestions to adopt the ESLint plugin for older apps and those who prefer not to use `create-next-app`

# Impact
We’ve encountered many NextJs sites that have impacted _core web metrics metrics_ because of some simple mistakes. 
Some Eg: 
- incorrectly loading 1P and 3P synchronously and/or with higher priority than core hydration scripts.
- Users loading CSS without preload, or not adding critical CSS before script tags.
- Using native `a` tag, causing full page refresh for NextJS apps.

The ESLint plugin will help catch a subset of these issues early on and also provide some baseline recommendations/best practices to write code in NextJS.
Another benefit will be offloading work form the `webpack-conformance-plugin`, which would make build time conformance checks faster.

# List of ESlint plugins (WIP)
- **Sync scripts** - Check for script tags being added without async/defer
All of core NextJS scripts are loaded async. Any user added script tag without a async/defer attribute takes priority over nextJS scripts and delays hydration.
- **No CSS tags** - Prevent users from loading CSS with <link> tags
We want users to import their CSS in recommended way using @import
NextJS preloads CSS and loads them with priority. CSS added with <link> tags runs the risk of being loaded at a different priority and increasing FCP metric
- **No <a> tag** - Warn users when they use an <a> tag instead of next/link
<a> tags cause full page reloads for SPA Next apps
- **Pollyfill.io checks** - Check if users using the polyfill.io api to load polyfills are re-loading polyfills that come built in with Next
- **Dynamic routes without `as` attribute** - Dynamic routes without the `as` attribute can lead to broken links that do not hydrate and cause full page reloads when clicked

_We've started work on these plugins and currently testing them with few Next apps. Also working to identify other ESLint rules that would help._

**Comment by ties-v** [comment]
Maybe it is also interesting to tackle some of the [common errors](https://github.com/zeit/next.js/tree/canary/errors) also during development by means of this plugin?

**Reply by janicklas-ralph** [reply]
That's a good idea.

**Comment by sergiodxa** [comment]
It will be great to also use this to warn about deprecated features, e.g. use getServerSideProps or getStaticProps instead of getInitialProps (yes, it's not actually deprecated but the general recommendation I believe is to use one of the new methods).

**Comment by Janpot** [comment]
> _Also working to identify other ESLint rules that would help._

Not sure if these categorize under "conformance", but some more lint ideas that have popped up before:
- [`no-anonymous-default-export`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-anonymous-default-export.md) for https://github.com/zeit/next.js/issues/12291
- [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks) for https://github.com/zeit/next.js/issues/12291#issuecomment-621139158
- [`no-unresolved`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-unresolved.md#casesensitive) for https://github.com/zeit/next.js/issues/11742

**Reply by prateekbh** [reply]
These are great @Janpot, we can probably use the already existing ones for these in a default config. Meanwhile continuing to create some of the Next specific rule living in the Lint plugin itself 👍🏻

**Comment by ties-v** [comment]
It seems that the Chrome team has already started, see #11986

**Reply by janicklas-ralph** [reply]
Yes. The PR #11986 was a follow up of this RFC.
